### ISSUE TITLE
Usage of multiple cores during report generation

### ISSUE THREAD

**Issue by LymanZerga** [danielpalme/ReportGenerator#126]
Hi,

we have recognized that the generator doesn't take advantage of all available cores. We have very powerful build agents where ReportGenerator is executed, but it seems that it doesn't scale.

32 Cores 1,5 hours
88 Cores almost the same 1,5 hours

Is it possible, to extend ReportGenerator by a new command argument specifing the amount of cores ReportGenerator should take care of during generation?

**Comment by danielpalme** [comment]
Hi,

I will consider this for the next release.
It won't work for every report type. But for HTML it should be possible.

Your execution time seems extremely long.
How many classes and lines of code does your code base have?

**Comment by LymanZerga** [comment]
More than one million lines of code...it's a huge project with many assemblies from different parties.

**Comment by danielpalme** [comment]
Could you please try this version:
https://www.nuget.org/packages/ReportGenerator/4.0.0-alpha2

I'm now using several cores to generate the class reports
Let me know if this works for you or if you have any issues.

How many hours/minutes does it take to generate the report now?

**Comment by LymanZerga** [comment]
Hi Daniel,

I will try out tomorrow. Do i have to use a special argument or do you implemented mutlicore usage by default?

**Comment by danielpalme** [comment]
Nothing special to do. Just replace the existing version.

**Comment by LymanZerga** [comment]
Thanks i‘ll keep you informed.

**Comment by danielpalme** [comment]
Thanks for testing.
_ReportGenerator_ first parses the XML file(s) generated by _OpenCover_.

The individual class files are only parsed when the report file for that class is generated.

In your case it seems that the parsing of the XML files generated by _OpenCover_ fails.
How big are these file(s)?

**Comment by LymanZerga** [comment]
About 80MB each file.

**Comment by danielpalme** [comment]
Strange that you don't have issues with the current version _ReportGenerator_, I did not change anything regarding the parsing of the XMLs.

How many XML files are passed to _ReportGenerator_?

Can you share one of those large reports by mail (danielpalme82@gmail.com)?
You can of course obfuscate your namespace and class names if you like?

If not, I have to create a large file myself.

**Comment by LymanZerga** [comment]
No we never had such an exception before.
It‘s about 500 files and each of it has something like 80mb.
Could you please create such a big file on your own because of security restrictions i am not allowed to share it.

**Comment by danielpalme** [comment]
OK, will create one myself.

**Comment by LymanZerga** [comment]
Ok if you need further informations let me know.

**Comment by danielpalme** [comment]
Will analyze this within the next days (maybe related to #100 )

**Comment by danielpalme** [comment]
Sorry for the delay, I was too busy.

Now I tried to reproduce the issue by generating some sample data.
I tried the following approaches:
* Parsing the same file in a loop
* Parsing different files with different modules and classes

The files had about ~100MB and memory usage always was between 150MB and 800MB.
Hard to figure out what's different on your system/input files.

**Comment by LymanZerga** [comment]
With what degree of parallizm did you execute your example?

**Comment by danielpalme** [comment]
1) The XML files from OpenCover are read in a simple loop (no parallizm):
https://github.com/danielpalme/ReportGenerator/blob/v4/src/ReportGenerator.Core/Parser/ParserFactory.cs#L37

2) The modules/classes within a single file are processed in parallel:
https://github.com/danielpalme/ReportGenerator/blob/v4/src/ReportGenerator.Core/Parser/OpenCoverParser.cs#L69
https://github.com/danielpalme/ReportGenerator/blob/v4/src/ReportGenerator.Core/Parser/OpenCoverParser.cs#L322

3) When all XML files have been read and processed, the HTML files are generated in parallel:
https://github.com/danielpalme/ReportGenerator/blob/v4/src/ReportGenerator.Core/Reporting/ReportGenerator.cs#L138

Your `OutOfMemoryException `is thrown in step 2, when one of the report files is processed.

**Comment by LymanZerga** [comment]
Hard to say where the memory exception occurs exactly and why. It seems that you to much XDocuments open during parsing.
I will look at the code on monday and try to make a proposal how to avoid such a memory exception.

regards Marcus

**Comment by LymanZerga** [comment]
In addition:
Can you try to replace Environment.ProcessorCount by 20 or 32 for your local run just to check if the memory exception occurs?

**Comment by danielpalme** [comment]
"Replacing  Environment.ProcessorCount by 20 or 32" will not have any impact.
This only applies to **step 3** ( see [comment](https://github.com/danielpalme/ReportGenerator/issues/126#issuecomment-388618559) above) but the memory issue is related to **step 2**

I removed the parallel processing during initialization in **step 2**. Perhaps you want to try this version:
https://www.nuget.org/packages/ReportGenerator/4.0.0-alpha5
If this does not help, we have to dig deeper. Perhaps _XDocument_ is a problem here.

**Comment by LymanZerga** [comment]
Sorry Daniel 

had no time to look at your code yet.
Will try your Version tomorrow.

 I think it would help to implement a producer consumer pattern where the Producer creates the parser and add them into a blocking collection. The consumer reads from this collection and proceed with step2 and step3. With that approch you will reduce the load memory as you only creates a bunch of parser instead all of them at once.

**Comment by danielpalme** [comment]
You are right, that's not perfect yet. I will change this and come back to you.
I don't think it makes sense to test the current version, just wait for an update.
I will fix this within the next days.

**Comment by LymanZerga** [comment]
Thanks

**Comment by danielpalme** [comment]
I looked at the code again and the workflow currently is:

1. First _OpenCover_ file gets parsed as _XDocument_ and is converted into a list of [Assemblies](https://github.com/danielpalme/ReportGenerator/blob/v4/src/ReportGenerator.Core/Parser/Analysis/Assembly.cs).
2. The next file gets parsed and the resulting [Assemblies](https://github.com/danielpalme/ReportGenerator/blob/v4/src/ReportGenerator.Core/Parser/Analysis/Assembly.cs) are merged with the result of step 1
3. Repeat **step 2** until all files are parsed

That means only **one** _XDocument_ is in memory at the same time.

When I profile the parsing process the memory is mainly used by _XAttribute_ and _XElement_ objects, but is always below 700 MB (when parsing 100 MB _OpenCover_ files):
![memory_usage](https://user-images.githubusercontent.com/859606/40280049-1194b59c-5c4e-11e8-9f96-37e474f9e1e2.png)

The only way I see to reduce memory usage is to replace _XDocument_ with [XmlReader
](https://msdn.microsoft.com/en-us/library/system.xml.xmlreader.aspx)

How much memory is used in your case when _ReportGenerator_ crashes?

**Comment by LymanZerga** [comment]
I have no easy way to analyze at which amount of memory the generator is crashing. I can run your new version and check if i can get a crash dump.

What if you convert the xdocument into custom data? We have also a custom that converts those xml coverage files in parallel and this does work without any problems. 
What if you run your new algorithm with lets say max degree of parallizm of 12 does it crash with a out of memory exception?

**Comment by LymanZerga** [comment]
I can run your first parallel version again on my local 20 core machine and make a dot trace dump file. Maybe this helps to find out where the memory is consumed?

**Comment by danielpalme** [comment]
I convert every XML file to a custom (minimal) object tree (a list of [Assemblies](https://github.com/danielpalme/ReportGenerator/blob/v4/src/ReportGenerator.Core/Parser/Analysis/Assembly.cs)).
I'm not able to force an OutOfMemoryException, even when I parse the XDocuments in parallel.
My sample files still get parsed this time with about 4 GB of memory usage instead of 700 MB.

Not sure what's different with your files.

A dump could be helpful and perhaps you can execute ReportGenerator with the Visual Studio profiler?
Or at least create a screenshot of the memory usage chart in TaskManager when it crashes?!

**Comment by LymanZerga** [comment]
I will do on thuesday or latest on wednesday. 
On monday with have holiday 😬

I try to do both a vs profiler output file and a dot trace output file.

**Comment by danielpalme** [comment]
Thanks!

**Comment by danielpalme** [comment]
You can use the latest version, it contains some performance improvements (memory usage is not changed):
https://www.nuget.org/packages/ReportGenerator/4.0.0-alpha7

**Comment by LymanZerga** [comment]
Hi Daniel,

 

today I tried the version 4.0.0-alpha7 and after ~25 minutes the reporter crashed. In the log file the last line only says that there are several errors occurred. I guess you must unwrap the AggregationException thrown by the .Net TPL and log each containing InvalidOperationException on their own.

 

I wanted to use the vs memory profiler to create a dump file but a have some problems during installation. We use the VS2015 Enterprise Edition but any time a want to start the profiler the VS want to start an installer. After that my pc is crashing each time :(  I have created a ticked for our local IT desk, hope they can provide a fix for me. 

 

Meanwhile I can offer you a dotTrace memory dump file, but I am pretty sure you can only open it with dotTrace from jetBrains. Would this be helpful for you? I could also provide a common dump file created by ProcessExplorer; problem is that I need to create it before the reporter is crashing. Assuming it crashes frequently after about ~20 minutes (locally in my machine) a can make a dump after 15 and 18 minutes, would this be helpful for you?

 

In addition I have one question to your overall parsing process. I assume it should work as follows:

 

Parallel foreach (file =>

{

     using (var  parser = CreateParserFor(file))

    {

        var htmlFile = parser.CreateHtmlOutput();

        htmlFile.Save();

    }

}, …);

 

But you said step 1 could not be parallized, could you explain this a bit in detail?

 

We create test impact files out of all those coverage files and we do all in parallel. We parsing the file with a XDocument instance (same as you), using LINQ to iterating over the XElements (same as you) and converting it to c# data object representations (in your case parser instances). Afterwards we writing a *.impact file for each xml file (in your case html file). 

For ~520 Files á 100MB (in sum 50GB) we need about 8 Minutes for that on a 12 Core agent (logical cores). We never got a OutOfMemoryException so far.

 

Hope this helps a bit. 

 

Regards Marcus

Von: Daniel Palme [mailto:notifications@github.com] 
Gesendet: Montag, 21. Mai 2018 21:04
An: danielpalme/ReportGenerator <ReportGenerator@noreply.github.com>
Cc: Lori <rollerfreak2@gmx.de>; Author <author@noreply.github.com>
Betreff: Re: [danielpalme/ReportGenerator] Usage of multiple cores during report generation (#126)

 

You can use the latest version, it contains some performance improvements (memory usage is not changed):
https://www.nuget.org/packages/ReportGenerator/4.0.0-alpha7

**Comment by danielpalme** [comment]
A dump file created by _ProcessExplorer_ would be a good start. This can be analyzed with _WinDbg_.

The whole parsing process is a bit more complex than in your case
Each XML file generated by _OpenCover_ is parsed as a _XDocument_.
Each parser produces a List<Assembly> (my internal data structure).

All those lists are merged into a single list.

The resulting list is rendered as HTML.

The first step is not run in parallel to minimize memory usage. Moreover the merging of the list is a rather complex process which cannot run in parallel.

I will add some better error logging soon.

**Comment by LymanZerga** [comment]
Hi Daniel,

 

sorry for the late response. I had to clarify some company restrictions meanwhile. Since a dump file contains data of the process and also data of our coverage files, I am not allowed to send you such a dump :(

 

But: I had started a report generator instance locally for round about 500 coverage files (each file ~100MB). I have observed that the memory rises and declines like a sawtooth (garbage collection). But like up to 2,8GB, back to 1GB,  up to 3GB, back to 1,2GB, up to 3,2GB and so on. After a while a received some out of memory exceptions like on our build agents. 

 

Did you tried to reproduce this memory issue with such an amount of 50GB coverage data? Maybe the assembly list could also be an issue why the memory exception happens!?

 

Any ideas how to proceed?

 

Von: Daniel Palme [mailto:notifications@github.com] 
Gesendet: Mittwoch, 23. Mai 2018 20:48
An: danielpalme/ReportGenerator <ReportGenerator@noreply.github.com>
Cc: Lori <rollerfreak2@gmx.de>; Author <author@noreply.github.com>
Betreff: Re: [danielpalme/ReportGenerator] Usage of multiple cores during report generation (#126)

 

A dump file created by ProcessExplorer would be a good start. This can be analyzed with WinDbg.

The whole parsing process is a bit more complex than in your case
Each XML file generated by OpenCover is parsed as a XDocument.
Each parser produces a List (my internal data structure).

All those lists are merged into a single list.

The resulting list is rendered as HTML.

The first step is not run in parallel to minimize memory usage. Moreover the merging of the list is a rather complex process which cannot run in parallel.

I will add some better error logging soon.

**Comment by danielpalme** [comment]
I also can confirm the sawtooth pattern. The spikes come from the XDocuments.
The assembly list also takes some memory, but I can't avoid that. And the sawtooth pattern shows, that is memory usage goes down after the XDocument is released.

I just published a new release:
https://www.nuget.org/packages/ReportGenerator/4.0.0-alpha8

Two major changes:

- Better error messages when parsing a files fails (Aggregate exceptions are unwrapped)
- Continue with next file instead of a complete failure (this means that a file can be missing in the analysis)

Let's see what the result on your 500 files is with that release.

Do those 500 files all cover the same assemblies and classes or are they completely different?
That information would help me to generate similar XML files to finally force an OutOfMemoryException on my machine.

**Comment by LymanZerga** [comment]
I will try this version on Monday. 

 

Regarding our coverage:

Our integration tests covering a lot more than the assemblies of our sub-project. But we call open cover to just listening to files/assemblies belonging to our sub-project.

 

Overall during our integration tests there are more than 200 assemblies loaded but (as mentioned above) we are just interested in a bunch of them (~15). The same for report generator, we are just interested in a coverage overview for our assemblies. I need to check if the coverage files really just containing source information of our assemblies. Will do on Monday.

 

Our 15 assemblies contains of about 5000 c# classes and ~700k lines of source code. Just to get an overview of what can be the difference between your run and our run.

 

Regards Marcus

 

Von: Daniel Palme [mailto:notifications@github.com] 
Gesendet: Freitag, 25. Mai 2018 22:44
An: danielpalme/ReportGenerator <ReportGenerator@noreply.github.com>
Cc: Lori <rollerfreak2@gmx.de>; Author <author@noreply.github.com>
Betreff: Re: [danielpalme/ReportGenerator] Usage of multiple cores during report generation (#126)

 

I also can confirm the sawtooth pattern. The spikes come from the XDocuments.
The assembly list also takes some memory, but I can't avoid that. And the sawtooth pattern shows, that is memory usage goes down after the XDocument is released.

I just published a new release:
https://www.nuget.org/packages/ReportGenerator/4.0.0-alpha8

Two major changes:

*	Better error messages when parsing a files fails (Aggregate exceptions are unwrapped)
*	Continue with next file instead of a complete failure (this means that a file can be missing in the analysis)

Let's see what the result on your 500 files is with that release.

Do those 500 files all cover the same assemblies and classes or are they completely different?
That information would help me to generate similar XML files to finally force an OutOfMemoryException on my machine.

**Comment by danielpalme** [comment]
Ok, the only way to figure this out is to create coverage files that match your setup.
In my tests v4 has a 25% better performance than v3.

So I have to figure out what's different in your situation. This will take some time.

**Comment by danielpalme** [comment]
I tried to force an `OutOfMemoryException` with different coverage files:

- Size: 150 to 350 MB 
- One assembly with many small/large classes, several assemblies with small/large classes
- Assemblies covering same assemblies/classes, Assemblies covering different assemblies/classes

In my tests no exception occurred.

I'm closing this issue for now as your coverage files seem to be different from my test setup.
I would be happy to help und to fix this issue, but without more information it is really hard.

Feel free to reopen if you can provide some data. 

- You could anonymize the namespaces, class and method names
- I can also sign a NDA
- We could use a temporary virtual machine on Azure or AWS where you place the coverage files

**Comment by LymanZerga** [comment]
Hi Daniel,

thanks for your response. I am still out of office but when you sign an NDA i should be allowed to give you the report data.

I come back to you when i am back in office (2 weeks)

Regards Marcus

**Comment by LymanZerga** [comment]
We will prepare a NDA for you, we just need to know:

- which data do you need? Coverage files and assemblies?
- email address --> if we need further data

regards Marcus

**Comment by danielpalme** [comment]
The coverage files are enough.
My email: danielpalme82@gmail.com

**Comment by danielpalme** [comment]
With the coverage files I was able to reproduce the issue and fix it.
You could try this new release:
https://www.nuget.org/packages/ReportGenerator/4.0.0-rc6

I did a quick performance comparison. In my setup the new 4.x version is significantly faster than the 3.x version:
![performance](https://user-images.githubusercontent.com/859606/46227197-bd980980-c35e-11e8-96db-a27c63c85531.png)

**Comment by LymanZerga** [comment]
Sounds Great Daniel! I will try your new Version on monday!

I‘ll keep you informed!

Von meinem iPad gesendet
